### PR TITLE
channeldb: write the CommitSig in CommitDiff in LN wire format

### PR DESCRIPTION
**Pull request by Roasbeef**
In this commit, we modify the way we write the CommitSIg in certain locations to match the way it's encoded on the wire. With this change, we'll now properly add a length prefix in front of the message, meaning we'll be able to read the extra data included (if present), and still be able to serialize content after this message.

We add a simple migration to migrate all the current pending commitments, along with a test. We opt to do the simple enclose-copy method for the migration, which results in a larger diff, but IMO an easier to understand migration test. 

This PR is needed for #3966  to pass the tests properly.

### CODE REVIEW

File: channeldb/channel.go [hunk 1]
```diff
@@ -1680,7 +1680,7 @@ func serializeCommitDiff(w io.Writer, diff *CommitDiff) error {
 		return err
 	}
 
-	if err := diff.CommitSig.Encode(w, 0); err != nil {
+	if err := WriteElements(w, diff.CommitSig); err != nil {
```
Comment (cfromknecht) [inline channeldb/channel.go]: be coupled with the last commit? o/w running just this commit could break your db

File: channeldb/migration14/enclosed_codec.go [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+	return fmt.Sprintf("Unknown type in %s: %T", e.method, e.element)
+}
+
+// WriteElement is a one-stop shop to write the big endian representation of
+// any element which is to be serialized for storage on disk. The passed
+// io.Writer should be backed by an appropriately sized byte slice, or be able
+// to dynamically expand to accommodate additional data.
+func WriteElement(w io.Writer, element interface{}) error {
+	switch e := element.(type) {
+	case keychain.KeyDescriptor:
```
Comment (cfromknecht) [inline channeldb/migration14/enclosed_codec.go]: non-blocking, but could trim unneeded types

File: channeldb/migration14/migration.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	openChanBucket := tx.Bucket(openChannelBucket)
+	if openChanBucket == nil {
+		return nil
+	}
+
+	return openChanBucket.ForEach(func(nodePub, v []byte) error {
+		// Ensure that this is a key the same size as a pubkey,
+		// and also that it leads directly to a bucket.
+		if len(nodePub) != 33 || v != nil {
+			return nil
```
Comment (cfromknecht) [inline channeldb/migration14/migration.go]: shouldn't it be an error if it doesn't? or are there other keys in the bucket?

File: channeldb/migration14/migration.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/coreos/bbolt"
+)
+
+// MigrateCommitDiff migrates the CommitDiff struct on-disk for all active
+// channels to store the CommitSig with a length prefix, just as we do on the
+// wire. This ensures we'll be able to store any extra data appended to the
+// message, yet still serialize items after the message as well on disk.
+func MigrateCommitDiff(tx *bbolt.Tx) error {
+	openChanBucket := tx.Bucket(openChannelBucket)
```
Comment (cfromknecht) [inline channeldb/migration14/migration.go]: do we also need to migrate the historical channels bucket?

File: channeldb/migration14/migration.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+				// If there's a value, it's not a bucket so
+				// ignore it.
+				if v != nil {
+					return nil
+				}
+
+				// Once we've found a valid channel bucket,
+				// we'll extract it from the node's chain
+				// bucket.
+				chanBucket := chainBucket.Bucket(chanPoint)
```
Comment (cfromknecht) [inline channeldb/migration14/migration.go]: missing nil check?

File: channeldb/migration14/migration.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	return openChanBucket.ForEach(func(nodePub, v []byte) error {
+		// Ensure that this is a key the same size as a pubkey,
+		// and also that it leads directly to a bucket.
+		if len(nodePub) != 33 || v != nil {
+			return nil
+		}
+
+		nodeChanBucket := openChanBucket.Bucket(nodePub)
+		if nodeChanBucket == nil {
+			return nil
```
Comment (cfromknecht) [inline channeldb/migration14/migration.go]: would also think this is an error

File: channeldb/migration14/migration.go (hunk 3 above)
Comment (Roasbeef) [reply channeldb/migration14/migration.go]: This was copied from our current logic where it served as more of a sanity check. Can add an error, could possibly surface issues with legacy DBs perhaps.

File: channeldb/migration14/migration.go (hunk 4 above)
Comment (Roasbeef) [reply channeldb/migration14/migration.go]: Thought about that, but I don't think so, since we don't also migrate the `commitDiffKey` for them as there are no pending commitments given the channel is now closed.

File: channeldb/migration14/enclosed_codec.go (hunk 2 above)
Comment (Roasbeef) [reply channeldb/migration14/enclosed_codec.go]: Removed types not needed for the migration.

File: channeldb/channel.go (hunk 1 above)
Comment (Roasbeef) [reply channeldb/channel.go]: Good idea, done.

File: channeldb/migration14/migration.go (hunk 5 above)
Comment (Roasbeef) [reply channeldb/migration14/migration.go]: If this is nil, then the user has issues with their database, as there's an invalid key at this level. Added the nil check nevertheless.

File: channeldb/migration14/migration.go (hunk 4 above)
Comment (cfromknecht) [reply channeldb/migration14/migration.go]: interesting, i was thinking we just copied over the state in-place, but if that's the case then perhaps not

**Comment by cfromknecht** [comment]
@Roasbeef linter is failing https://travis-ci.org/github/lightningnetwork/lnd/jobs/664666715#L737, latest version looking pretty close tho 👍

File: channeldb/migration14/migration.go (hunk 4 above)
Comment (Roasbeef) [reply channeldb/migration14/migration.go]: We _do_ copy over the state in-place, but that only the "core" channel state: https://github.com/lightningnetwork/lnd/blob/master/channeldb/channel.go#L2608 

The `commitDiffKey` is only updated here: https://github.com/lightningnetwork/lnd/blob/master/channeldb/channel.go#L1955 

Earlier in the transaction when we close the channels, we call [this method which clears out the commit diff state](https://github.com/lightningnetwork/lnd/blob/master/channeldb/channel.go#L2558).

File: channeldb/migration14/enclosed_codec.go [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd/lnwire"
```
Comment (joostjager) [inline channeldb/migration14/enclosed_codec.go]: One drive-by comment: here we are still relying on packages that may change and break the migration. These are basic types, so it is unlikely, but the migration isn't fully self-contained like for example a sql migration script would be or migration 13.

File: channeldb/migration14/legacy_decoding.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+			&d.ClosedCircuitKeys[i].HtlcID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &d, nil
+}
+
+func serializeHtlcsLegacy(b io.Writer, htlcs ...HTLC) error {
```
Comment (halseth) [inline channeldb/migration14/legacy_decoding.go]: do we really need legacy serialization methods?

File: channeldb/migration14/migration.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+				// nothing to migrate.
+				commitDiffBytes := chanBucket.Get(commitDiffKey)
+				if commitDiffBytes == nil {
+					return nil
+				}
+
+				// Now that we have the commit diff in the
+				// _old_ encoding, we'll write it back to disk
+				// using the new encoding which has a length
+				// prefix in front of the CommitSig.
```
Comment (halseth) [inline channeldb/migration14/migration.go]: We've discussed before the pros of doing more low-level migrations. Can this be done here as well? Copy the bytes up to the message index, add a length byte, then copy the remaining bytes. Would minimize the dependencies needed and the amount of code to copy.

File: channeldb/migration14/legacy_decoding.go (hunk 8 above)
Comment (Roasbeef) [reply channeldb/migration14/legacy_decoding.go]: Yep, we need to insert it into the database in the legacy format for the migration. See the migration code.

File: channeldb/migration14/migration.go (hunk 9 above)
Comment (Roasbeef) [reply channeldb/migration14/migration.go]: I'm not too worried about deps or the amount of code. The code is really just dead code we put in there to gain confidence in the migration itself. The deps can be frozen using modules if we choose as well. Personally I can't really verify the low-level migrations during review without _re-generating_ all the bytes myself, so just more manual work, vs being able to read _deterministic_ code.

Comment (Roasbeef) [reply channeldb/migration14/migration.go]: > Copy the bytes up to the message index, add a length byte, then copy the remaining bytes

The amount of bytes preceding the message can't easily be deduced as we write the channel commitment before the diff, and the commitment contains elements with a variable size (the amount of HTLCs). The commitment itself isn't written with any sort of length prefix either.

**Comment by Roasbeef** [comment]
Just pushed an update to this PR to make it now work properly with #3966. If you merge these two PRs into each other, all the unit tests should now pass. The changes in this new version are:
  * Modify `codec.go` to always include a length prefix. This changes the way we need to write the unsigned unacked updates to disk, so that's migrated as well. 
  * Ensure we write the network results with that same length prefix. 

The "migration" commits themselves are still small, as `channeldb/codec.go` does all the heavy lifting.

File: channeldb/migration14/enclosed_codec.go (hunk 7 above)
Comment (Roasbeef) [reply channeldb/migration14/enclosed_codec.go]: The main package here that impacts the migration is `lnwire`, specifically the way we write wire messages to disk. I don't think this will be changing any time soon. If they attributes or types of any of the imported packages change, then if they materially affect the migration, this PR should no longer compile.

Comment (joostjager) [reply channeldb/migration14/enclosed_codec.go]: Especially the wire format is something that isn't set in stone, but of course it depends on what our support period is for this migration. Also a future change may be detected because the system doesn't build anymore, but that does mean that more copying needs to happen at that point. Touching an old migration to make it build again is a risk. I favor the fully isolated low-level approach.

**Comment by joostjager** [comment]
I would reconsider merging this for 0.10. Touching the persistent channel state is risky regardless of how big or small the migration is. If we need to, it is safer to do it at the beginning of the cycle. As far as I can see, this pr and #3966 are only refactorings and don't add new functionality or fixes, so why take the risk?

File: channeldb/migration14/legacy_decoding.go (hunk 8 above)
Comment (halseth) [reply channeldb/migration14/legacy_decoding.go]: For the migration we only need to _deserialize_ the old format as we are never gonna write it again. I see the legacy serialization is used for the tests only, so could be beneficial to move the methods there if they are needed, but it is non-blocking.

File: channeldb/migration14/migration.go (hunk 9 above)
Comment (halseth) [reply channeldb/migration14/migration.go]: > > Copy the bytes up to the message index, add a length byte, then copy the remaining bytes
> 
> The amount of bytes preceding the message can't easily be deduced as we write the channel commitment before the diff, and the commitment contains elements with a variable size (the amount of HTLCs). The commitment itself isn't written with any sort of length prefix either.

Mhm, agree that that makes it tricky.

File: channeldb/migration14/migration.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+					return err
+				}
+
+				var b bytes.Buffer
+				err = serializeCommitDiff(&b, commitDiff)
+				if err != nil {
+					return err
+				}
+
+				err = chanBucket.Put(
```
Comment (halseth) [inline channeldb/migration14/migration.go]: We should be careful here, as the bbolt docs state that 
> ForEach executes a function for each key/value pair in a bucket. [...] The provided function must not modify the bucket; this will result in undefined behavior.

Not sure if modifying a subbucket of the bucket counts as modifyinng the bucket itself.

File: channeldb/migration14/migration.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+		var newResultBuf bytes.Buffer
+		err = serializeNetworkResult(
+			&newResultBuf, oldResult,
+		)
+		if err != nil {
+			return err
+		}
+
+		return networkResults.Put(k, newResultBuf.Bytes())
```
Comment (halseth) [inline channeldb/migration14/migration.go]: same comment about not modifying bucket during `ForEach`. This one definitely looks unsafe.

File: channeldb/migration14/enclosed_codec.go (hunk 7 above)
Comment (Roasbeef) [reply channeldb/migration14/enclosed_codec.go]: Copying is essentially zero cost, it's all effectively dead code anyway. As mentioned offline, making `migration14` into a module should resolve any issues with the referenced types/packages changing under us (so it should always build given the deps are static).

File: channeldb/codec.go [hunk 12]
```diff
@@ -178,7 +179,17 @@ func WriteElement(w io.Writer, element interface{}) error {
...
+		if _, err := lnwire.WriteMessage(&msgBuf, e, 0); err != nil {
+			return err
+		}
+
+		msgLen := uint16(len(msgBuf.Bytes()))
+		if err := WriteElements(w, msgLen); err != nil {
+			return err
+		}
+
+		if _, err := w.Write(msgBuf.Bytes()); err != nil {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/codec.go]: By wire I mean as it's written to our "socket". That "socket" then wraps the wire message in an encrypted transport. The wire message itself doesn't include the length prefix, the transport actually does. 

The two options for length encoding are either varint, or just a `uint16`. I've gone with the `uint16` here.

File: channeldb/migration14/migration.go (hunk 10 above)
Comment (Roasbeef) [reply channeldb/migration14/migration.go]: Hmm yeah I had that same thought here when writing this section myself...updating to just do multiple passes so we're on the safe side.

**Comment by Roasbeef** [comment]
Updated to now migrate the channel close summaries. We don't need to migrate the forwarding package or the discovery message store as those wire messages are the sole value being stored in the key. The channel summary migration itself is forwarding looking as we don't store anything after the message, but this change allows us to append new fields to the close summary in the future. 

Please see the fixup commits as I had to modify some of the legacy decode/encode methods to bypass `Read/WriteElements` and use `lnwire.Read/WriteMessage` directly as the imported codec includes the new serialization for wire messages with the length prefix.

File: channeldb/migration14/legacy_decoding.go [hunk 13]
```diff
@@ -42,12 +44,22 @@ func deserializeLogUpdatesLegacy(r io.Reader) ([]LogUpdate, error) {
...
 		if err != nil {
+			fmt.Println("No more for me")
 			return nil, err
 		}
+
+		updateMsg, err := lnwire.ReadMessage(r, 0)
+		if err != nil {
+			fmt.Println("tried to read everything")
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (joostjager) [reply channeldb/migration14/legacy_decoding.go]: In any case, I wouldn't squash this with the copy/paste commit. At least make a PURE MOVE commit (not necessarily building) and do all required changes in a separate commit. Otherwise it becomes hard to review.

File: channeldb/codec.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply channeldb/codec.go]: Ok, so it happens to be the same, but we could just as well have used any of the other ways to encode the length, like the length encoding in our tlv format? Or even encode the message in single record tlv format, so that we can add more fields in the future without relying on EOF?

File: channeldb/migration14/migration.go [hunk 14]
```diff
@@ -59,112 +65,180 @@ func MigrateDatabaseWireMessages(tx kvdb.RwTx) error {
...
+	for _, closedChan := range closedChans {
+		oldSummary, err := deserializeCloseChannelSummaryLegacy(
+			bytes.NewReader(closedChan.summaryBytes),
+		)
+		if err != nil {
+			return err
+		}
+
+		var newSummaryBytes bytes.Buffer
+		err = serializeChannelCloseSummary(
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/migration14/migration.go]: > y. But why do the migration of the close summary now then? Why take the risk if we don't need it now and possibly never? 

You brought it up, and it falls under the same case as the other instances fixed in this PR. IMO we might as well bundle these up so we can have one less migration in the future.

File: channeldb/codec.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/codec.go]: On the wire level, we still need to rely on the EOF, or do incremental byte accounting to make sure we don't read more than we need to. Here in the database, we can use any of those formats. Doing this migration now means we can actually use TLV within the DB for any extensions in the future.

File: channeldb/migration14/migration.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/migration14/migration.go]: > For the single-message keys, you could also argue that we should migrate them in case we add something more to it in the future.

Now that we're aware of the issue, a migration isn't the only way to resolve this issue for the single-value use case. They can either use another key or some other option.

[the remark this answers is not in the document]
Comment (joostjager) [reply channeldb/migration14/migration.go]: I brought them up as something to check, but didn't mean to say that they should be migrated if it isn't strictly needed.

Migration isn't the only way to resolve this issue for the single-value use case, but neither is it the only way to resolve the issue for multi-valued use cases. Depending on db structure, another key could be added there too.

File: channeldb/codec.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply channeldb/codec.go]: Suppose you want to add a tlv block at the end (after the message that is migrated in this pr) in a future pr, then you still need either a migration or an EOF check to see if the tlv block is there or not?

What I suggested is to introduce the tlv block already in this PR. Not converting all the existing fields, but just replace the message by a single-record tlv block.

Existing format: "MESSAGE"
This pr: 7 | "MESSAGE" (prepend length)
Suggestion: 0 | 7 | "MESSAGE" (make it tlv already now with the message being type 0, so that more data can be added later)

This then does need to be moved to the very end of the bbolt value, otherwise another length field is needed.

But yes, it is just a suggestion. Can also see that we don't want to extend it like that now.

**Comment by Roasbeef** [comment]
Rebased and squashed. Ready for a fresh review as a lot of context has been lost over the past 2 months.

File: channeldb/migration14/legacy_decoding.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/migration14/legacy_decoding.go:55]: Moved to a new commit.

File: channeldb/channel.go [hunk 15]
```diff
@@ -1785,12 +1785,12 @@ func deserializeLogUpdates(r io.Reader) ([]LogUpdate, error) {
 	return logUpdates, nil
 }
 
-func serializeCommitDiff(w io.Writer, diff *CommitDiff) error {
+func serializeCommitDiff(w io.Writer, diff *CommitDiff) error { // nolint: dupl
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/channel.go:1824]: I'd rather include this single line in this commit, than create a distinct commit for it. It doesn't add much noise to this commit.

**Comment by Roasbeef** [comment]
> With the recent insights that we gained about migrations, their risks and scalability, shouldn't this PR be converted into an on-the-fly migration?

My current determination is that the bucket items touched in this PR aren't "high risk", as the migration doesn't touch buckets with a large amount of content (like invoices or payments), and the migration also doesn't migrate over an entire key set. 

The contents touch in this migration are:
  * the set of open channels _that have a pending commitment a migration time_ 
 * the set of _closed_ channels 
 * the set of channels to be announced where we haven't received ann sigs from the remote party 

Compared to the prior issues that we saw in the migration of 10s of thousands and even hundreds of thousands of keys, the locations referenced above typically have an order of magnitude less data. The other issues with the prior migration was the create of so many sub-buckets and keys. This migration doesn't create any new keys or sub-buckets, it instead replaces the current value stored for a sub-set of keys.

File: channeldb/migration14/migration_test.go [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+			networkResults, err := tx.CreateTopLevelBucket(
+				networkResultStoreBucketKey,
+			)
+			if err != nil {
+				return err
+			}
+
+			return networkResults.Put(
+				netResultKey, netResBuf.Bytes(),
+			)
```
Comment (joostjager) [inline channeldb/migration14/migration_test.go:231]: This setup code seems to bear a risk that the db isn't constructed exactly as it was in older versions. Ideally it would be constructed using high-level methods, but that probably is even more copy/paste...

**Comment by Roasbeef** [comment]
> I've liked the low level approach of migration 13 with test snapshots, but as we discussed before, that may not work as well for the more complicated data structures in this pr.

When I went to edit that migration to attempt to optimize things a bit more, I found I didn't really know how to edit the tests if I wanted to, so make sure that any optimizations were properly tested. In the end, it's a tradeoff at the end of the day. IMO copying code is _very_ minor, and is well worth it if users are able to modify/understand the migration more easily. 

> One question is what the best timing is for this PR to merge. Could this interfere with our investigation into current channel state machine instabilities?

I think we've held off on this long enough, with the PR really changing little in the mean time. This does indeed touch the commitment diff, so like any other migration, we'll want to ensure we're able to run this normally on an existing node.

File: channeldb/migration14/legacy_decoding.go [hunk 17]
```diff
@@ -43,11 +43,19 @@ func deserializeLogUpdatesLegacy(r io.Reader) ([]LogUpdate, error) {
...
 	for i := 0; i < int(numUpdates); i++ {
 		err := ReadElements(r,
-			&logUpdates[i].LogIndex, &logUpdates[i].UpdateMsg,
+			&logUpdates[i].LogIndex,
 		)
 		if err != nil {
 			return nil, err
 		}
+
+		updateMsg, err := lnwire.ReadMessage(r, 0)
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/migration14/legacy_decoding.go:52]: Yeah turns out this was totally unnecessary. Removed the commit all together.

File: channeldb/migration14/migration_test.go (hunk 16 above)
Comment (Roasbeef) [reply channeldb/migration14/migration_test.go:231]: Well we only care about these set items and not the entire DB right? The attempt here by copying over the old code is to ensure the items on disk are the same as if they were written in the "past".

File: channeldb/migration14/enclosed_codec.go [hunk 18]
```diff
@@ -0,0 +1,375 @@
+package migration14
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/migration14/enclosed_codec.go:1]: Well one can review it, just either they need to `diff` commands to check them against the versions of the code checked in before this commit.

**Comment by Roasbeef** [comment]
Pushed up a new version with a newer migration number: 15.

**Comment by Roasbeef** [comment]
Also I ended up removing that last commit to create a module. As I realized it isn't possible for a sub-module and a greater module to depend on distinct versions of the same packages. Only if we split up the migrations into a standalone binary would something like this be possible.

File: channeldb/migration15/migration.go [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+// currently store wire messages without length prefixes. This includes the
+// CommitDiff struct, ChannelCloseSummary, LogUpdates, and also the
+// networkResult struct as well.
+func MigrateDatabaseWireMessages(tx kvdb.RwTx) error {
+	openChanBucket := tx.ReadWriteBucket(openChannelBucket)
+	if openChanBucket == nil {
+		return nil
+	}
+
+	// The migration will proceed in three phases: we'll need to update any
```
Comment (joostjager) [inline channeldb/migration15/migration.go:20]: Phases could be distinct functions

File: channeldb/migration15/new_encoding.go [hunk 20]
```diff
@@ -0,0 +1,431 @@
...
+
+	return nil
+}
+
+func serializeCommitDiff(w io.Writer, diff *CommitDiff) error { // nolint: dupl
+	if err := serializeChanCommit(w, &diff.Commitment); err != nil {
+		return err
+	}
+
+	if err := WriteElements(w, diff.CommitSig); err != nil {
```
Comment (joostjager) [inline channeldb/migration15/new_encoding.go:99]: This change is already the new style serialize, making this commit not a pure copy. How to find out whether more changes were made in this commit? The only way I know of that worked for review in other PRs is to make a pure copy commit that does not make any changes at all and then apply the changes on top of that in a second commit.

File: channeldb/codec.go [hunk 21]
```diff
@@ -389,7 +400,13 @@ func ReadElement(r io.Reader, element interface{}) error {
 		*e = bytes
 
 	case *lnwire.Message:
-		msg, err := lnwire.ReadMessage(r, 0)
+		var msgLen uint16
+		if err := ReadElement(r, &msgLen); err != nil {
```
Comment (cfromknecht) [inline channeldb/codec.go:404]: do we know that `CommitSig` is the only `lnwire.Message` that isn't length-prefixed?

These are all the messages I found passing through this codec in `channeldb` unit tests
```
DECODING CommitSig
DECODING UpdateAddHTLC
DECODING UpdateFailHTLC
DECODING UpdateFulfillHTLC
ENCODING CommitSig
ENCODING UpdateAddHTLC
ENCODING UpdateFailHTLC
ENCODING UpdateFulfillHTLC
```

Comment (Roasbeef) [reply channeldb/codec.go:404]: I think one other message is the `ChannelReestablish` that we store within the channel close summary. In the pst I think I had portions of this in this PR, but it's been a bit since I've rebased the PR that adds the actual change in the `lnwire` logic. 

I'm gonna rebase that on top of this, and run the tests there, which in the past have let me catch other issues in this base migration.

Comment (Roasbeef) [reply channeldb/codec.go:404]: Oh I think those are the `LogUpdates` in the greater commit diff.

File: channeldb/migration15/migration.go [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+		// At this point, we have the channel bucket now, so we'll
+		// check to see if this channel has a pending commitment or
+		// not.
+		commitDiffBytes := chanBucket.Get(commitDiffKey)
+		if commitDiffBytes != nil {
+			// Now that we have the commit diff in the _old_
+			// encoding, we'll write it back to disk using the new
+			// encoding which has a length prefix in front of the
+			// CommitSig.
+			commitDiff, err := deserializeCommitDiffLegacy(
```
Comment (Roasbeef) [inline channeldb/migration15/migration.go:106]: Need to test this, but I think it's ok @cfromknecht. We'll read things out using the old format here, which includes the log updates without the length prefix. Down below, we'll then write things out using the new format that includes the length prefix for each message. 

The chan reest migration is still missing though from what I can tell.

Comment (Roasbeef) [reply channeldb/migration15/migration.go:106]: We migrated the `CommitSig` message in isolation because we used the raw `Encode/Decode` methods on it rather than `lnwire.[Read/Write]Message`.

**Comment by Roasbeef** [comment]
Pushed an update as `migration17` (no longer 15).

File: channeldb/codec.go (hunk 21 above)
Comment (Roasbeef) [reply channeldb/codec.go:409]: `ChannelReestablish` should be handled when we migrate the closed channel summaries using the same construct:
   * read out using the old non-length prefixed version
   * write out using the new length prefixed version

**Comment by Roasbeef** [comment]
So I ran this PR in isolation against an existing node. Everything checked out. I then ran the combined branch with #3966 which shone light on an existing issue (brought up earlier I believe):
  * When the migration is run with #3966, it uses the new `lnwire` methods that cause it to attempt to continue reading until an EOF is hit.
   * This causes the migration to _read_ out too much data during the translation process, which garbles the wire messages. 

If the prior sub-module idea worked like we wanted it to, then we could've pinned this dependnacy, and avoided this issue. 

In order to avoid needing to copy over pretty much _all_ of `lnwire`, I think we can instead add an intermediate commit to create a slimmed down `lnwire.ReadMessage` that manually reads in each field for the set of messages we care about. This method would be used in the migration, and would let us avoid having to copy over all the structs. 

Thoughts?

**Comment by Roasbeef** [comment]
Alternatively, I can instead modify #3966 to actually use the protocol version (`pver`) argument of `lnwire.[Read/Write]Message`. As the `pver` is already threaded down all the way to the individual `Encode/Decode` methods, we could then make the "TLV read until EOF" only apply to `pver=1`. The migration would use `pver=0` when reading and `pver=1` when writing. Everything else would then be updated to use `pver=1`. 

This doesn't add any new copied code and uses an existing extension/backwards-compat feature that has been unused until now. I'm leaning towards this approach myself.

**Comment by Roasbeef** [comment]
For future migrations, I think this could've been caught if we add a "post-condition" to the dry run mode. In this case, it would've tried to read out a commit diff or some other field, hit this error, then rolled everything back.

**Comment by Roasbeef** [comment]
Need to also add a migration for the forwarding packages since they use `codec.ReadElements`.

**Comment by cfromknecht** [comment]
> Alternatively, I can instead modify #3966 to actually use the protocol version (pver) argument of lnwire.[Read/Write]Message. As the pver is already threaded down all the way to the individual Encode/Decode methods, we could then make the "TLV read until EOF" only apply to pver=1. The migration would use pver=0 when reading and pver=1 when writing. Everything else would then be updated to use pver=1.

This indeed sounds like the best approach. If we go this route, maybe it makes sense to have a global constant to prevent us from accidentally forgetting what version we are on. Currently it's just a phantom argument that people don't think about, so we would want to make sure that people don't forget to use the right one in the future.

**Comment by Roasbeef** [comment]
> Alternatively, I can instead modify #3966 to actually use the protocol version (pver) argument of lnwire.[Read/Write]Message. 

Updated the other PR to implement this. This change ties these two PRs closer to each other once more, as #3966 won't pass the unit tests without this PR. I want to combine them, but then the resulting diff would be rather large.

**Comment by Roasbeef** [comment]
> Need to also add a migration for the forwarding packages since they use codec.ReadElements.

Correction, the forwarding packages are OK since they write out to a value that only has the log updates and nothing else. The issue was with `lnwire`, which is now resolved in #3966.

**Comment by joostjager** [comment]
>Updated the other PR to implement this. This change ties these two PRs closer to each other once more, as #3966 won't pass the unit tests without this PR. I want to combine them, but then the resulting diff would be rather large.

And how about combine this PR with #3966 excluding the addition of `ExtraOpaqueData` in all messages?

File: channeldb/migration18/migration.go [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+	var closedChans []closedChan
+
+	closedChanBucket := tx.ReadWriteBucket(closedChannelBucket)
+	if closedChannelBucket == nil {
+		return fmt.Errorf("no closed channels found")
+	}
+	err = closedChanBucket.ForEach(func(k, v []byte) error {
+		closedChans = append(closedChans, closedChan{
+			chanKey:      k,
+			summaryBytes: v,
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/migration18/migration.go:167]: The bytes aren't used outside the transaction, so I think so.

**Comment by Roasbeef** [comment]
This PR is going away now (in favor of the combined version), but I pushed up a rebased version for posterity.
